### PR TITLE
Add the awaited attribute to CoroutineMock.

### PR DESCRIPTION
**Pull request by Kentzo**
Refs #64.

### CODE REVIEW

**Comment by Kentzo** [comment]
@Martiusweb despite the incompatibility with 3.4 (which I will fix) what do you think?

File: test/test_mock.py [hunk 1]
```diff
@@ -283,6 +283,23 @@ def test_called_CoroutineMock_returns_MagicMock(self):
         mock = asynctest.mock.CoroutineMock()
         self.assertIsInstance(run_coroutine(mock()), asynctest.mock.MagicMock)
 
+    def test_awaited_CoroutineMock_sets_awaited(self):
```
Comment (Martiusweb) [inline test/test_mock.py]: Can you add more tests:
* when result is awaitable?
* when awaiting the result raises an exception?

File: asynctest/mock.py [hunk 2]
```diff
@@ -405,6 +405,9 @@ class CoroutineMock(Mock):
     :class:`unittest.mock.Mock` object: the wrapped object may have methods
     defined as coroutine functions.
     """
+    awaited = unittest.mock._delegating_property('awaited')
```
Comment (Martiusweb) [inline asynctest/mock.py]: I have a concern about the name of this property (see my comment in the PR)

**Comment by Martiusweb** [comment]
Yes, this is something that should be implemented.
We can add the method assert_awaited(), but it can be done in an other PR.
Do you have examples of cases where an Event is useful for awaited (rather than a simple boolean)?
It can be a good idea but it can be confusing knowning that unittest.mock.Mock.called is a boolean: for instance, someone will probably do something like:

    self.assertTrue(coro_mock.awaited)

which will always be ok. I think it should be renamed in awaited_event (or both attributes can be provided, awaited as a boolean, awaited_event as an Event object).

About the exception: the way you implemented it is right. Since you check that the returned result is awaitable, the exception will be raised as the result of the awaited call, not because the object is not awaitable. Thus, I think that we can safely say that the CoroutineMock has been awaited even if an exception is raised.

**Comment by Kentzo** [comment]
I cannot share the entire project, but the use case is for tests whose behavior require to do something at very specific steps of execution flow.

E.g. when a test involves 2 coroutines (client and server) and you want to send some data after server authorizes client. In that test one of server's methods is wrapped with a mock and then test awaits that wrapped mock to be awaited. That is a signal that test can proceed.

> It can be a good idea but it can be confusing knowning that unittest.mock.Mock.called is a boolean

What if we replace asyncio.Event with a custom subclass:

```python
class Event(asyncio.Event):
    def __bool__(self):
        return self.is_set()
```

> I think it should be renamed in awaited_event

The reason I did it this way was to make it closer to `called` / `call_count`.

**Comment by Martiusweb** [comment]
I recently merged autospec support but it seems that I forgot a few things (handling the mock attributes/delegated properties correctly).
In fact, I didn't use ``unittest.mock._delegating_property()`` for is_coroutine and a few other private properties of the mocks, but I should have.

**Comment by Kentzo** [comment]
Do you know how to address it? I was going to take a look at this next.

**Comment by Martiusweb** [comment]
Your last commit is the right way to fix it. I thought it was caused by how _mock_add_spec is overridden, but I forgot that adding methods/public attributes requires adding it to create_autospec().

I'll add a bit of documentation and tests for the autospec case and merge it.

**Comment by Martiusweb** [comment]
Since I don't have the right to edit your PR, I pushed my changes in the branch GreatFruitOmsk-coroutine-wait:
I rebased master (I fixed the documentation generation which was broken since it used values from setup.py that we moved in setup.cfg), and added 3 commits.
* 80b799f and 85e454f are cosmetic/doc changes
* 92bb286 implements tests for the methods with create_autospec and fix them.

Can you review them so we can merge it all? It will be part of the 0.12 release, that I will release once I'm ready to merge #73 and #74.

**Comment by Kentzo** [comment]
- I think there is no reason to reference "old" `yield from` syntax
- Perhaps the `awaited` Event-like object should be extended with the `wait_next` method: important if test wants to distinguish 1st, 2nd etc awaits

File: test/test_mock.py [hunk 3]
```diff
@@ -351,16 +349,48 @@ def test_awaited_from_autospec_mock(self):
...
         run_coroutine(mock.a_coroutine())
 
         mock.a_coroutine.assert_awaited()
         self.assertTrue(mock.a_coroutine.awaited)
-        self.assertTrue(mock.a_coroutine.awaited.is_set())
         self.assertEqual(1, mock.a_coroutine.await_count)
 
+    def test_awaited_wait(self):
+        loop = asyncio.new_event_loop()
+        with replace_loop(loop):
```
Comment (Martiusweb) [inline test/test_mock.py]: I don't think it is worth changing run_coroutine, which is just a very small helper.
I'd simply use the normal APIs of asyncio and unittest:

    loop = asyncio.new_event_loop()
    self.addCleanup(loop.close)

    mock = asynctest.mock.CoroutineMock()
    t = asyncio.ensure_future(mock.awaited.wait())
    loop.run_until_complete(mock())
    (...)

File: asynctest/mock.py [hunk 4]
```diff
@@ -383,12 +383,62 @@ class MagicMock(AsyncMagicMixin, unittest.mock.MagicMock,
...
+        self._mock = mock
+        self._condition = asyncio.Condition()
+
+    @asyncio.coroutine
+    def wait(self, skip=0):
+        """
+        Wait for await to be called at least ``skip`` number of times.
+        """
+        def predicate(mock):
+            return mock.await_count > skip
```
[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: There is inconsistency indeed, but I like the `skip` argument as it appears more to the point: you wait for a call skipping a given number of other calls. Perhaps I could rephrase the doc better.

File: asynctest/mock.py [hunk 5]
```diff
@@ -383,12 +383,62 @@ class MagicMock(AsyncMagicMixin, unittest.mock.MagicMock,
...
+        """
+        await_count = self._mock.await_count
+
+        def predicate(mock):
+            return mock.await_count > await_count + skip
+
+        return (yield from self.wait_for(predicate))
+
+    @asyncio.coroutine
+    def wait_for(self, predicate):
```
[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: @Martiusweb I think github's parser ate the prefix, what did you mean `_`?

File: test/test_mock.py (hunk 3 above)
Comment (Kentzo) [reply test/test_mock.py]: @Martiusweb That's not enough. The new loop is also need to be set and unset (like run_coroutine do) before mock is created, otherwise `asyncio.Condition` will capture the current loop, not the loop that is running it.

File: asynctest/mock.py [hunk 6]
```diff
@@ -383,12 +383,62 @@ class MagicMock(AsyncMagicMixin, unittest.mock.MagicMock,
...
+
+    @asyncio.coroutine
+    def wait_next(self, skip=0):
+        """
+        Wait for await to be called at least ``skip + 1`` more times.
+        """
+        await_count = self._mock.await_count
+
+        def predicate(mock):
+            return mock.await_count > await_count + skip
```
[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: I find the `skip` name more to the point: wait for await skipping # awaits.

Please take a look at the rephrased docstrings. If they are still hard to comprehend, I'll change the name.

File: test/test_mock.py (hunk 3 above)
Comment (Martiusweb) [reply test/test_mock.py]: You're right. We can keep it that way for now, or move these tests in a class extending ``asynctest.TestCase``, since one of its purpose is to handle the loop for each test case.

File: asynctest/mock.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Martiusweb) [reply asynctest/mock.py]: Yes, it sounds consistent like this:

    mock.awaited.wait() # wait the first await
    mock.awaited.wait(skip=1) # don't resume for the first await

cool :)

File: asynctest/mock.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Martiusweb) [reply asynctest/mock.py]: Interresting use case. I didn't think of something like this.

I think that we can add this example in the documentation, and tell that the predicate is only checked when the mock is awaited. I don't want users to have the feeling that it can magically awake after a random event.

[the remark this answers is not in the document]
Comment (Martiusweb) [reply asynctest/mock.py]: In fact, I was thinking about this:

    def predicate(mock):
        return mock.await_count > 0 and m.call_args == (("foo", ), {})

    t1 = asyncio.ensure_future(mock("foo"))
    t2 = mock("bar")  # not yet scheduled
    mock.awaited.wait_for(predicate)  # will hang forever: mock.call_args is now based on the last call.

The problem is that the predicate is based on the mock (ie coroutine function), not the coroutine that will be awaited itself.

I'm affraid that this will be confusing as it's not possible to wait_for() a given instance of the coroutine function to be awaited. Maybe we should keep this private and let users write their own logic when they want to do complex things (one can use the ``wraps`` argument to customize th behavior of the mock).

[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: You should use `call_args_list`.

[the remark this answers is not in the document]
Comment (Martiusweb) [reply asynctest/mock.py]: It doesn't solve the issue of knowing which instance you're awaiting:

    def predicate_wait_for_bar(mock):
        return mock.await_count > 0 and (("bar", ), {}) in m.call_args_list

    t1 = asyncio.ensure_future(mock("foo"))
    t2 = mock("bar")  # not yet scheduled
    mock.awaited.wait_for(predicate_wait_for_bar)  # predicate will be true

[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: You should find an index of the call in `call_args_list` and compare it against `await_count`.

[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: This won’t be 100% accurate (you can await in different order) but could be sufficient for some tests.

[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: In other hand for those tests `skip` would be sufficient as well. Perhaps we should add `await_args` / `await_args_list`?

[the remark this answers is not in the document]
Comment (Martiusweb) [reply asynctest/mock.py]: I think that the use case for checking which exact call has been awaited makes sense, but I'm not conviced with generic/public wait_for() method yet.

I think that we should merge this PR with a private _wait_for() as it's already very useful like this. Then we can work on a new PR to implement wait_call() and wait_next_call() or add optional parameters matching in wait() and wait_next().

[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: When I was working on that, ability to await a specific call was only one of use cases I thought of. Another was an ability to await an external consition like setting mock’s attribute to a specific value. Or more general: any modification of mock’s state related to await. I also thought of ability to await external conditions, but couldn’t think of a practical example.

I would opt for having `wait_for` public as it seems in line with the naming of async primitives. People who would confuse it would probably have problems with `wait` and `wait_next` as well. This an advanced method, but it is possible to mark it as such with docs.
 
What do you say if we keep it public via naming, but exclude from the documentation for now?

[the remark this answers is not in the document]
Comment (Kentzo) [reply asynctest/mock.py]: Please check https://github.com/GreatFruitOmsk/asynctest/commit/3b22978c031ee23a222546ff3e36538a0cb3edaf, it introduces an independent list of "await args" and among other address the issue in your example for `wait_for`.

**Comment by Martiusweb** [comment]
Hi,

Sorry for the delay.
I'll do a few cosmetic changes (typos, etc) and merge it. I'll also add a loop argument to CoroutineMock (in an other PR) to be able to control which loop is used in _AwaitedEvent.

I was thinking of something else:
currently a user is notified once the coroutine that is awaited returned, but doesn't it make more sense to notify that the coroutine has been awaited as soon as it starts to be executed?

Basically, it would be something like this:

    @asyncio.coroutine
    def proxy():
        try:
            yield from (asyncio.sleep(0))
        finally:
            _mock_self.await_count += 1
            yield from _mock_self.awaited._notify()

        return (yield from result)

**Comment by Kentzo** [comment]
Please take a look at the commit I linked. It is not part of this PR, but I’d like to include it.

> I was thinking of something else

In that case you likely don’t care about executing the coroutine and plain side_effect would do.

**Comment by Martiusweb** [comment]
I added a bunch of comments, but this is very cool, let's add this commit in the PR!

I changed a few cosmetic things and rebased the PR on top of master (I pushed a change to fix something in python 3.7).

Once the PR is updated with the changes of https://github.com/Martiusweb/asynctest/tree/GreatFruitOmsk-coroutine-wait and the commit for await args is ready, we can merge.

Thank you very much!

**Comment by Martiusweb** [comment]
There still work to do to fix Python 3.7 with PYTHONASYNCIODEBUG=1, I'll ignore this for this PR and work on it later.

**Comment by Martiusweb** [comment]
Finally done!

Thank you very much for this awesome contribution and your patience.

**Comment by Kentzo** [comment]
Just pushed a follow up ;)

Please let me know if anything needs to be changed for other PRs I opened.
